nemo/collections/tts/models/tacotron2.py

```diff
@@ -0,0 +1,282 @@
+# Copyright (c) 2020, NVIDIA CORPORATION.  All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from dataclasses import dataclass
+from typing import Dict, List, Optional
+
+import torch
+from omegaconf import MISSING, DictConfig, OmegaConf, open_dict
+from torch import nn
+from torch.nn.functional import pad
+
+from nemo.collections.tts.helpers.helpers import get_mask_from_lengths, tacotron2_log_to_tb_func
+from nemo.core.classes import Loss, ModelPT, typecheck
+from nemo.core.neural_types.elements import (
+    AudioSignal,
+    EmbeddedTextType,
+    LengthsType,
+    LogitsType,
+    LossType,
+    MelSpectrogramType,
+    SequenceToSequenceAlignmentType,
+)
+from nemo.core.neural_types.neural_type import NeuralType
+from nemo.utils import logging
+from nemo.utils.decorators import experimental
+
+
+@dataclass
+class PreprocessorParams:
+    pad_value: float = MISSING
+
+
+@dataclass
+class Preprocessor:
+    cls: str = MISSING
+    params: PreprocessorParams = PreprocessorParams()
+
+
+@dataclass
+class Tacotron2Config:
+    preprocessor: Preprocessor = Preprocessor()
+    encoder: Dict = MISSING
+    decoder: Dict = MISSING
+    postnet: Dict = MISSING
+    labels: List = MISSING
+    train_ds: Optional[Dict] = None
+    validation_ds: Optional[Dict] = None
+
+
+class Tacotron2Loss(Loss):
```

Comment:
this class should go to nemo/collections/tts/losses/* or to nemo/collections/common/losses/* and also get some docstring

Reply:
Why move it? This loss is only ever going to be used in this model. It is so redundant that I have to define it as it's own separate class rather than a function of Tacotron2Model.

Reply:
With the new function-level override of typecheck(), it doesn't even need to inherit from Typing to use NeMo's typing. `Tacotron2Loss` doesn't have an init nor does it use `self`, it's function.

Reply:
For consistency. We put all losses other losses under collections/collection_name/losses/* .  why make exception here?

Reply:
Moved.
I maintain that `collections/collection_name/losses/*` shouldn't exist apart from losses inside common. NLP's losses is empty and ASR only has CTC that is only used in CTC models so it should be moved there and losses folder should be removed.

Reply:
Let me qualify my above statement. If a loss is only ever going to be used in a single model, it should not be in the losses folder. And if the losses folder is empty, it should be deleted.

Reply:
It is incorrect that the class odes not need to implement Typing. All of the actual tests lie inside the Typing class, and the decorator simply dispatches these checks. This is by design to support graph level checks. The decor will raise Runtime error if placed on a class that doesn't implement Typing.